### PR TITLE
Add ability to restrict room history.

### PR DESCRIPTION
**Pull request by erikjohnston**
This adds the following restrictions to which events users/servers can see:

> If `m.room.history_visibility` event has `visibility` set to `after_join` then:
> - A user can only see that event if they were joined at that event.
> - A server can only see an unredacted version of that event if one of its users were joined at the time.

TODO:
- [x] Update names
- [x] Implement `invited` option
- [x] Set default power level for the event.

### CODE REVIEW

**Comment by erikjohnston** [comment]
Suggestions for better names for the following very welcome:
- `m.room.history_visibility`
- `visibility`
- `after_join`

We should also define a value for the current default, e.g. `joined`? Which is distinct from something like `all`/`public` that allows even those that haven't joined to see history.

**Comment by ara4n** [comment]
i suggest:
 * event name: m.room.history_visibility
 * event key: history_visibility # for consistency with the event name
 * event value:
   * public: anyone can see it, whether they're in the room or not (starting from the point where this is set)
   * shared: currently joined members can see all history (starting from the point where the event is set)
   * private: currently joined members can see history from the point they joined.

What is the powerlevel name you need to set this?

**Comment by erikjohnston** [comment]
From IRL : We also want the ability to allow currently *invited* members to see history. Proposal is to change `private` => `invited` and `joined`.

**Comment by erikjohnston** [comment]
We also need to set a default power level to change this, probably 100?

**Comment by erikjohnston** [comment]
I've implemented the following protection for `m.room.history_visibility` events:
- New rooms will now have power level required to change it set to `100`
- Existing rooms will fall back to the level set by `state_default`, which by default is `50`.

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -222,6 +224,52 @@ def log_failure(f):
...
+                if visibility in ["invited", "joined"]:
+                    for ev in state.values():
+                        if ev.type != EventTypes.Member:
+                            continue
+                        try:
+                            domain = UserID.from_string(ev.state_key).domain
+                        except:
+                            continue
+
+                        if domain != server_name:
```
Comment (NegativeMjark) [inline synapse/handlers/federation.py]: What is this check doing?

Comment (erikjohnston) [inline synapse/handlers/federation.py]: This is checking if the given server is in the room by looping through all the membership events to find users for that server.

Could probably do with one of those "comment" things I keep hearing about.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**